mandible/metadata_mapper/source_provider.py

```diff
@@ -0,0 +1,140 @@
+import logging
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Optional, Type, TypeVar
+
+from .context import ContextValue
+from .format import FORMAT_REGISTRY
+from .source import SOURCE_REGISTRY, FileSource, Source
+from .storage import STORAGE_REGISTRY
+
+log = logging.getLogger(__name__)
+
+T = TypeVar("T")
+
+REGISTRY_TYPE_MAP = {
+    "Format": FORMAT_REGISTRY,
+    "Storage": STORAGE_REGISTRY,
+    "Source": SOURCE_REGISTRY,
+}
+
+
+class SourceProviderError(Exception):
+    pass
+
+
+class SourceProvider(ABC):
+    @abstractmethod
+    def get_sources(self) -> Dict[str, Source]:
```

Comment:
Do we want to add a return type when we are just passing in the method?

Reply:
It's an abstract method so we want to have type annotations so that subclasses will know what value to return when they implement the method. 

The `pass` statement is never actually executed since the `ABC` system will raise an exception if you try to instantiate a class without overriding the abstract method implementation. You could put anything there, but the standard convention is to use `pass`.